python/src/spark_rapids_ml/umap.py

```diff
@@ -946,45 +966,110 @@ def _fit(self, dataset: DataFrame) -> "UMAPModel":
+            sc: pyspark.SparkContext,
+            input: Union[np.ndarray, Dict[str, np.ndarray]],
+            BROADCAST_LIMIT: int,
+        ) -> Union[
+            List[pyspark.broadcast.Broadcast],
+            Dict[str, List[pyspark.broadcast.Broadcast]],
+        ]:
+            """
+            Broadcast the input array, chunking it into smaller arrays if it exceeds the broadcast limit.
+            If the input is a dictionary of arrays (i.e. CSR), chunk and broadcast each array in the dictionary.
+            """
+            if isinstance(input, np.ndarray):
+                total_bytes = input.nbytes
+                if total_bytes <= BROADCAST_LIMIT:
+                    return [sc.broadcast(input)]
+
+                rows_per_chunk = BROADCAST_LIMIT // (total_bytes // input.shape[0])
+                num_chunks = (input.shape[0] + rows_per_chunk - 1) // rows_per_chunk
+                return [
+                    sc.broadcast(input[i * rows_per_chunk : (i + 1) * rows_per_chunk])
+                    for i in range(num_chunks)
+                ]
+
+            elif isinstance(input, dict):
+                total_bytes = sum(map(lambda arr: arr.nbytes, input.values()))
+                if total_bytes <= BROADCAST_LIMIT:
+                    return {k: [sc.broadcast(arr)] for k, arr in input.items()}
+
+                rows_per_chunk = BROADCAST_LIMIT // (
+                    total_bytes // input["data"].shape[0]
+                )
+                num_chunks = (
+                    input["data"].shape[0] + rows_per_chunk - 1
+                ) // rows_per_chunk
+                # indptr is chunked differently because it has less rows; split to match num_chunks
+                indptr_rows_per_chunk = (
+                    input["indptr"].shape[0] + (num_chunks - 1)
+                ) // num_chunks
+                return {
+                    "indices": [
+                        sc.broadcast(
+                            input["indices"][
+                                i * rows_per_chunk : (i + 1) * rows_per_chunk
+                            ]
+                        )
+                        for i in range(num_chunks)
+                    ],
+                    "indptr": [
+                        sc.broadcast(
+                            input["indptr"][
+                                i
+                                * indptr_rows_per_chunk : (i + 1)
+                                * indptr_rows_per_chunk
+                            ]
+                        )
+                        for i in range(num_chunks)
+                    ],
+                    "data": [
+                        sc.broadcast(
+                            input["data"][i * rows_per_chunk : (i + 1) * rows_per_chunk]
+                        )
+                        for i in range(num_chunks)
+                    ],
+                }  # NOTE: CSR chunks are not independently meaningful; do not use until recombined
```

Comment:
Nit: we should be able to combine the duplicate code with a for loop like "for key in ["indices", "indptr", "data"]".

Reply:
Wrote it this way since indptr requires a different chunking index. indices and data code is indeed repetitive but this might just be clearer.